packages/kernel/src/kernel.ts

```diff
@@ -102,6 +135,15 @@ export abstract class BaseKernel implements IKernel {
       case 'history_request':
         await this._historyRequest(msg);
         break;
+      case 'comm_open':
+        await this._comm_manager.comm_open(msg);
```

Comment:
Wondering whether we should adopt the same pattern as for the other messages, by having methods on the `IKernel` interface? 

https://github.com/jtpio/jupyterlite/blob/97bbe1a0e8ec31108d905c3b7752666a01634851/packages/kernel/src/tokens.ts#L45

For example with `commOpen`, `commMsg` and `commClose`.

This could allow other kernels to completely change the implementation of the comm manager if they want, or just forward messages in the case of the Python kernel reusing the existing comm implementation.

Reply:
Yeah, as i mentioned, i'm still figuring out the architecture. the joke was, of course, first try to sneak into how one _might_ get this into pyolite, realizing that was way crazy, and then unwrapping the actual flow of messages through the system to get to _something_ that works.

This would of course need to move into its own package, but no reason not to have the API in better shape...

Reply:
> Now I am wondering if it's feasible and if it makes sense to reuse @bollwyvl 's JS implementation of the comms instead, we would need to expose those comms to the "Python process".

In theory probably yes (or hopefully yes), if the JS implementations follow what ipykernel expects. This could be passed as a custom Python package from JavaScript: https://pyodide.org/en/stable/usage/faq.html#how-do-i-create-custom-python-packages-from-javascript

Reply:
Right, I think we win if these two PRs can be harmonized, but prettied up, moved to some new packages....

```
packages/
  kernel/
    kernel.ts        # gets the comm(Open|Msg|Close)Msg handlers
  js-kernel/
    kernel.ts        # has new CommManager({kernel: this})
  py-kernel/src/
    worker.ts        # also needs CommManager in here...
    kernel.ts        # ... injects it here e.g. `import js; (await js.import(${COMM_URL})).CommManager`
    py/.../
      patches.py     # sys.modules hacks, as on #145
                     # ...and the shim to do-async everything from #141 js
  comms/src/
    comm.ts
    comm-manager.ts
    providers/
      memory.ts
      # TODO: localForage.ts
      # TODO: yjs.ts
```

In `patches.py`, there are probably some more pointy bits... At the `pyolite`/`import js` boundary we need it to pretty much be dumb JSON types and `Array`s, but it should be feasible.

Reply:
I'll hoist this back to #18